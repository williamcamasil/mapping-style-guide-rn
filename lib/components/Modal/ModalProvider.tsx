import React, {
  useState, ReactElement, useCallback, useMemo, useRef, Fragment, ComponentType, useContext, useEffect,
} from 'react';

import { ModalPropsType } from '.';

type ShowModalCallbackPropsType<P extends ModalPropsType > = Omit<P, 'visible' | 'isFocused' | 'modalKey' | 'theme'>;

export type ModalReturnType = {
  dismiss: () => void;
};

type ModalContextType = {
  showModal: (
    ModalComponent: ComponentType<ModalPropsType>,
    props?: ShowModalCallbackPropsType<ModalPropsType>
  ) => ModalReturnType;
  dismissAllModals: () => void;
};

const ModalContext = React.createContext<ModalContextType>({} as any);

type RenderedElementType = {
  modalKey: string;
  element: ReactElement<ModalPropsType>;
  notifyDismiss: () => void;
  notifyDismissEnd: () => void;
};

type ModalProviderPropsType = {
  children: React.ReactNode;
  /**
   * Tempo (em milissegundos) de tolerância entre a abertura de duas modais do mesmo tipo.
   *
   * Usado para evitar que o usuário abra várias vezes a mesma modal ao clicar
   * rapidamente no mesmo botão.
   *
   * Em ambiente de teste, você pode querer desativar esse comportamento, pois
   * em casos de testes os disparos de modais acontecem várias vezes repetidamente.
   *
   * Para desativar, basta passar zero.
   */
  throttleTimeout?: number;
};

const ModalProvider: React.FC<ModalProviderPropsType> = ({
  children, throttleTimeout = 500,
}) => {
  const lastModalShowRef = useRef(0);
  const lastElementTypeRef = useRef<ReactElement['type']>();
  const [renderedElements, setRenderedElements] = useState<RenderedElementType[]>([]);

  const lastRenderedElementsRef = useRef<RenderedElementType[]>([]);

  /**
   * Dispara os eventos de dismiss e dismissEnd de acordo com as mudanças
   * de estado das modais.
   */
  useEffect(() => {
    lastRenderedElementsRef.current.forEach(lastElement => {
      const element = renderedElements.find(el => el.modalKey === lastElement.modalKey);

      if (!element) {
        lastElement.notifyDismissEnd();
        return;
      }

      if (!element.element.props.visible && lastElement.element.props.visible) {
        lastElement.notifyDismiss();
      }
    });

    lastRenderedElementsRef.current = renderedElements;
  }, [renderedElements]);

  const currentRenderedElement = renderedElements[renderedElements.length - 1];

  const dismissAllModals = useCallback(() => {
    setRenderedElements(elements => (
      elements.map(element => {
        if (element.element.props.visible) {
          return {
            ...element,
            element: React.cloneElement(element.element, { visible: false }),
          };
        }

        return element;
      })
    ));
  }, []);

  const unmountModal = useCallback((modalKey: string) => {
    setRenderedElements(elements => {
      const elementIndex = elements.findIndex(element => element.modalKey === modalKey);

      const copyElements = [...elements];

      copyElements.splice(elementIndex, 1);

      return copyElements;
    });
  }, []);

  const dismissModal = useCallback((modalKey: string) => {
    setRenderedElements(elements => elements.map(element => {
      if (element.modalKey === modalKey && element.element.props.visible) {
        return {
          ...element,
          element: React.cloneElement(element.element, { visible: false }),
        };
      }

      return element;
    }));
  }, []);

  const shouldThrottle = useCallback((element: ReactElement) => {
    try {
      if (!throttleTimeout) return false;

      if (element.type !== lastElementTypeRef.current) return false;

      const timeSinceLastShow = Date.now() - lastModalShowRef.current;

      if (timeSinceLastShow > throttleTimeout) return false;

      return true;
    } finally {
      lastElementTypeRef.current = element.type;
      lastModalShowRef.current = Date.now();
    }
  }, [throttleTimeout]);

  const showModal = useCallback(<Props extends ModalPropsType>(
    ModalComponent: ComponentType<Props>,
    props?: ShowModalCallbackPropsType<Props>,
  ) => {
    const modalKey = `modal-element-${lastModalShowRef.current}${Date.now()}`;

    const result = {
      dismiss: () => dismissModal(modalKey),
    };

    setRenderedElements(elements => {
      // eslint-disable-next-line react-perf/jsx-no-new-function-as-prop
      const handleDismissModal = () => dismissModal(modalKey);

      // eslint-disable-next-line react-perf/jsx-no-new-function-as-prop
      const handleDismissEndModal = () => unmountModal(modalKey);

      const element = (
        <ModalComponent
          {...props as any}
          visible
          onDismiss={handleDismissModal}
          onDismissEnd={handleDismissEndModal}
        />
      );

      if (shouldThrottle(element)) {
        return elements;
      }

      const renderedElement: RenderedElementType = {
        modalKey,
        element,
        notifyDismiss: () => {
          props?.onDismiss?.();
        },
        notifyDismissEnd: () => {
          props?.onDismissEnd?.();
        },
      };

      return [...elements, renderedElement];
    });

    return result;
  }, [dismissModal, shouldThrottle, unmountModal]);

  const providerValue = useMemo(() => ({
    showModal,
    dismissAllModals,
  }), [
    showModal,
    dismissAllModals,
  ]);

  const renderModalElement = (item: RenderedElementType) => {
    const clonedElement = React.cloneElement(item.element, {
      isFocused: currentRenderedElement?.modalKey === item.modalKey,
      modalKey: item.modalKey,
    });

    return (
      <Fragment key={item.modalKey}>
        {clonedElement}
      </Fragment>
    );
  };

  return (
    <ModalContext.Provider value={providerValue}>
      {children}
      {renderedElements.map(renderModalElement)}
    </ModalContext.Provider>
  );
};

export type ShowModalType = <Props extends ModalPropsType>(
  ModalComponent: ComponentType<Props>,
  props?: ShowModalCallbackPropsType<Props>
) => ModalReturnType;

export function useModal() {
  const { showModal } = useContext(ModalContext);
  return showModal as ShowModalType;
}

export function useDismissAllModals() {
  const { dismissAllModals } = useContext(ModalContext);
  return dismissAllModals;
}

export default ModalProvider;
