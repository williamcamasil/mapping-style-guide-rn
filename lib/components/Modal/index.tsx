
import React, {
  useState, useMemo, useCallback, useRef, useEffect, PropsWithChildren,
} from 'react';
import {
  StyleSheet, Animated, BackHandler, Easing,
  SafeAreaView, View, StyleProp, ViewStyle,
  Platform,
} from 'react-native';

import { useDimensions, useViewStyles } from '../../hooks';

export type ModalPropsType = PropsWithChildren & {
  testID?: string;
  visible: boolean;
  isFocused: boolean;
  modalKey?: string;
  onDismiss?: () => any;
  dismissible?: boolean;
  children?: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  overlayStyle?: StyleProp<ViewStyle>;
  onDismissEnd?: () => any;
  withSafeAreaIOS?: boolean;
  skipStatusBarAndroid?: boolean;
  animationDuration?: number;
  analyticsModalName?: string;
};

const styles = StyleSheet.create({
  portal: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'transparent',
  },
  safeArea: {
    flex: 1,
  },
  overlay: {
    flex: 1,
  },
  modal: {
    flex: 1,
  },
});

const Modal = ({
  visible,
  onDismiss,
  dismissible = true,
  children,
  style,
  overlayStyle,
  onDismissEnd,
  withSafeAreaIOS = true,
  skipStatusBarAndroid = false,
  animationDuration = 250,
  isFocused,
  testID = undefined,
  analyticsModalName,
  modalKey,
}: ModalPropsType) => {
  const $visible = useRef(false);
  const [rendered, setRendered] = useState(visible);

  const { window, statusBar } = useDimensions();
  const windowHeight = Math.max(window.width, window.height);

  const overlayOpacity = useRef(new Animated.Value(0));

  const modalTranslateY = useMemo(
    () => new Animated.Value(windowHeight),
    [windowHeight],
  );

  const showModal = useCallback(() => {
    setRendered(true);
    Animated.parallel([
      Animated.timing(modalTranslateY, {
        toValue: 0,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
      Animated.timing(overlayOpacity.current, {
        toValue: 1,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
    ]).start();
  }, [animationDuration, modalTranslateY]);

  const hideModal = useCallback(() => {
    Animated.parallel([
      Animated.timing(modalTranslateY, {
        toValue: windowHeight,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
      Animated.timing(overlayOpacity.current, {
        toValue: 0,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
    ]).start(() => {
      setRendered(false);
      onDismissEnd?.();
    });
  }, [modalTranslateY, windowHeight, animationDuration, overlayOpacity, onDismissEnd]);

  useEffect(() => {
    if ($visible.current === visible) return;

    $visible.current = visible;

    if (visible) {
      showModal();
    } else {
      hideModal();
    }
  }, [hideModal, showModal, visible]);

  useEffect(() => {
    if (visible) {
      const handleBack = () => {
        if (dismissible) {
          onDismiss?.();
        }
        return true;
      };

      const listener = BackHandler.addEventListener('hardwareBackPress', handleBack);

      return () => {
        listener.remove();
      };
    }
    return undefined;
  }, [visible, dismissible, onDismiss]);

  const overlayStyles = useViewStyles(() => [
    styles.overlay,
    overlayStyle,
    {
      opacity: overlayOpacity.current,
    },
  ], [overlayOpacity, overlayStyle]);

  const modalStyles = useViewStyles(() => [
    styles.modal,
    style,
    {
      marginTop: Platform.OS === 'android' && skipStatusBarAndroid ? statusBar.height : undefined,
      transform: [{
        translateY: modalTranslateY,
      }],
    },
  ], [modalTranslateY, skipStatusBarAndroid, statusBar.height, style]);

  if (!rendered) return null;

  const WrapperView = withSafeAreaIOS ? SafeAreaView : View;

  return (
    <View testID={testID} style={styles.portal} pointerEvents="box-none">
      <WrapperView testID="modal-safeArea" pointerEvents="box-none" style={styles.safeArea}>
        <Animated.View
          testID="modal-overlay"
          accessibilityViewIsModal
          accessibilityLiveRegion="polite"
          style={overlayStyles}
          onAccessibilityEscape={onDismiss}
        >
          <Animated.View
            testID="modal-content"
            style={modalStyles}
          >
            {children}
          </Animated.View>
        </Animated.View>
      </WrapperView>
    </View>
  );
};

export default Modal;
