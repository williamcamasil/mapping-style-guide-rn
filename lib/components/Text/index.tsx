import React from 'react';
import {
  StyleProp, StyleSheet, Text as RNText, TextProps as RNTextProp, TextStyle,
} from 'react-native';

import { useTextStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import {
  ColorPaletesType, FontWeightType, LineHeightType, Typography,
} from '../../tokens';
import {
  getColorStyle, getVariantStyle, getWeightStyle, TextVariantType,
} from './utils';

export type { TextVariantType } from './utils';

export type TextPropTypes = RNTextProp & {
  theme: AppThemeType;
  style?: StyleProp<TextStyle>;
  color?: ColorPaletesType;
  variant?: TextVariantType;
  weight?: FontWeightType;
  lineHeight?: LineHeightType;
};

const styles = StyleSheet.create({
  text: {
    letterSpacing: Typography.letterSpaces.default,
    flexShrink: 1,
  },
});

const Text: React.FC<TextPropTypes> = ({
  children,
  color,
  theme,
  style,
  variant = 'body',
  weight,
  lineHeight,
  allowFontScaling = false,
  ...others
}) => {

  const textStyles = useTextStyles(() => [
    styles.text,
    getColorStyle(theme, color ?? 'neutralGray600'),
    getVariantStyle(theme, variant, lineHeight),
    getWeightStyle(theme, weight),
    style,
  ], [color, lineHeight, style, theme, variant, weight]);

  return (
    <RNText
      {...others}
      allowFontScaling={allowFontScaling}
      style={textStyles}
    >
      {children}
    </RNText>
  );
};

export default withTheme(Text);
