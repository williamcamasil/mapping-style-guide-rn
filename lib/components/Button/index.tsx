import React, { ComponentType } from 'react';
import {
  StyleProp, StyleSheet, TouchableOpacity, TouchableWithoutFeedbackProps, View, ViewStyle,
} from 'react-native';

import { SvgProps } from 'react-native-svg';

import { useViewStyles } from '../../hooks/useStyles';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import Loading from '../Loading';
import Spacer from '../Spacer';
import Text, { TextPropTypes } from '../Text';
import {
  ButtonSizeType,
  ButtonVariantType,
  getSizeStyle,
  getTextColor,
  getVariantStyle,
  getStylesToBlock,
} from './utils';

export type ButtonPropsType<IconProps extends SvgProps = SvgProps> = {
  children?: TextPropTypes['children'];
  onPress: TouchableWithoutFeedbackProps['onPress'];
  style?: StyleProp<ViewStyle>;
  disabled?: TouchableWithoutFeedbackProps['disabled'];
  theme: AppThemeType;
  variant?: ButtonVariantType;
  size?: ButtonSizeType;
  loading?: boolean;
  testID?: string;
  Icon?: ComponentType<IconProps>;
  block?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
  textColorVariant?: ButtonVariantType;
};

const styles = StyleSheet.create({
  containerIconAndText: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

const Button: React.FC<ButtonPropsType> = ({
  children,
  style,
  onPress,
  disabled,
  theme,
  variant,
  size,
  loading,
  testID,
  Icon,
  block,
  containerStyle,
  textColorVariant,
}) => {
  const hasText = Boolean(children);

  const buttonStyles = useViewStyles(
    () => [
      {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: theme.borders.radius.medium,
      },
      getVariantStyle(theme, variant, disabled),
      getSizeStyle(theme, size, hasText),
      style,
    ],
    [disabled, hasText, size, style, theme, variant],
  );

  const containerStyles = useViewStyles(() => [getStylesToBlock(block), containerStyle], [block, containerStyle]);

  const textColor = getTextColor(textColorVariant ? textColorVariant : variant, disabled);

  const renderText = () => {
    if (!hasText) return null;
    return (
      <Text
        testID="text-with-icon"
        variant={size === 'small' ? 'bodySmall' : 'body'}
        weight="bold"
        lineHeight="medium"
        color={textColor}
      >
        {children}
      </Text>
    );
  };

  const renderContent = () => {
    if (loading) {
      return <Loading color={textColor} />;
    }

    if (Icon) {
      return (
        <View testID="container-icon" style={styles.containerIconAndText}>
          <Icon width={32} height={32} color={theme.colors[textColor]} />
          {children ? <Spacer size={theme.spacings.sXXS} /> : null}
          {renderText()}
        </View>
      );
    }

    return renderText();
  };

  return (
    <View testID="container-button" style={containerStyles}>
      <TouchableOpacity
        testID={testID}
        style={buttonStyles}
        onPress={onPress}
        disabled={disabled || loading}
        activeOpacity={theme.opacities.intense}
      >
        {renderContent()}
      </TouchableOpacity>
    </View>
  );
};

Button.defaultProps = {
  disabled: false,
  variant: 'containedPrimary',
  size: 'medium',
  block: true,
};

export default withTheme(Button);
