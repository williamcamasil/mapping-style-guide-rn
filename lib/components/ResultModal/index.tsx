import React from 'react';
import {
  StyleSheet, View,
} from 'react-native';

import AnimatedLottieView from 'lottie-react-native';
import { useDidMount } from 'mapping-context-rn';

import { Lotties } from '../../assets';
import { useViewStyles } from '../../hooks';
import { AppThemeType, withTheme } from '../../theme';
import Container from '../Container';
import Modal, { ModalPropsType } from '../Modal';
import Spacer from '../Spacer';
import Text from '../Text';

type ResultModalPropsType = ModalPropsType & {
  text: string;
  theme: AppThemeType;
  timeToCloseInMilliseconds?: number;
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    textAlign: 'center',
  },
});

const ResultModal: React.FC<ResultModalPropsType> = ({
  text, theme, timeToCloseInMilliseconds, onDismiss, ...others
}) => {
  const viewLottieStyles = useViewStyles(() => [
    {
      width: theme.spacings.lGiant,
      height: theme.spacings.lGiant,
    },
  ], [theme]);

  useDidMount(() => {
    const interval = setTimeout(() => {
      onDismiss?.();
    }, timeToCloseInMilliseconds);
    return () => clearTimeout(interval);
  });

  return (
    <Modal
      onDismiss={onDismiss}
      testID="result-modal"
      {...others}
    >
      <Container padding={theme.spacings.sLarge} style={styles.container}>
        <View style={viewLottieStyles}>
          <AnimatedLottieView
            source={Lotties.Check}
            autoPlay
          />
        </View>
        <Spacer size={theme.spacings.sXXS} />
        <Text
          variant="headingMedium"
          color="neutralGray700"
          weight="bold"
          lineHeight="medium"
          style={styles.text}
        >
          {text}
        </Text>
      </Container>
    </Modal>
  );
};

ResultModal.defaultProps = {
  timeToCloseInMilliseconds: 2000,
};

export default withTheme(ResultModal);
