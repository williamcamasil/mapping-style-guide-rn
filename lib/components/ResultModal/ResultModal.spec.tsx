import React from 'react';

import {
  render, waitFor, fireEvent,
} from '@testing-library/react-native';

import ResultModal from '.';
import { ThemeProvider } from '../../theme';
import Button from '../Button';
import Modal, { ModalPropsType } from '../Modal';
import ModalProvider, { useModal } from '../Modal/ModalProvider';
import Text from '../Text';

describe('ResultModal snapshot', () => {
  it('default', () => {
    const tree = render(
      <ResultModal
        isFocused
        visible
        text="text"
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('testing if have a message text', () => {
    const tree = render(
      <ResultModal
        isFocused
        visible
        text="Message text"
      />,
    );

    expect(tree.getByText('Message text')).toBeTruthy();
  });

  it('testing if modal closed', async () => {
    const handleDismiss = jest.fn();
    const result = render(
      <ResultModal
        isFocused
        visible
        onDismiss={handleDismiss}
        text="Message text"
        timeToCloseInMilliseconds={5}
      />,
    );

    expect(result.getByText('Message text')).toBeTruthy();

    await waitFor(() => {
      expect(handleDismiss).toBeCalled();
    });
  });

});

describe('ResultModal events', () => {
  it('should open another modal on dismiss', async () => {

    const OtherModal = (props: ModalPropsType) => (
      <Modal {...props}>
        <Text testID="other-modal-text">Hello</Text>
      </Modal>
    );

    const FakeScreen = () => {
      const show = useModal();

      const openAnotherModal = () => {
        show(OtherModal);
      };

      const openResultModal = () => {
        show(ResultModal, {
          text: 'Hello',
          onDismiss: openAnotherModal,
          timeToCloseInMilliseconds: 1,
        });
      };

      return (
        <Button
          testID="open-modal-button"
          onPress={openResultModal}
        >
          Open result modal
        </Button>
      );
    };

    const result = render(
      <ThemeProvider>
        <ModalProvider>
          <FakeScreen />
        </ModalProvider>
      </ThemeProvider>,
    );

    const openModalButton = result.getByTestId('open-modal-button');

    fireEvent.press(openModalButton);

    result.getByTestId('result-modal');

    await result.findByTestId('other-modal-text');
  });

});
