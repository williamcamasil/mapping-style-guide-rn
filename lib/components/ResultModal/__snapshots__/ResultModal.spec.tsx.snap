// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultModal snapshot default 1`] = `
<View
  pointerEvents="box-none"
  style={
    {
      "backgroundColor": "transparent",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
  testID="result-modal"
>
  <RCTSafeAreaView
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
    testID="modal-safeArea"
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      style={
        {
          "flex": 1,
          "opacity": 0,
        }
      }
      testID="modal-overlay"
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "marginTop": undefined,
            "transform": [
              {
                "translateY": 1334,
              },
            ],
          }
        }
        testID="modal-content"
      >
        <View
          pointerEvents="auto"
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "flex": 1,
              "flexDirection": undefined,
              "justifyContent": "center",
              "padding": 28,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "paddingVertical": undefined,
            }
          }
        >
          <View
            style={
              {
                "height": 200,
                "width": 200,
              }
            }
          >
            <View
              style={
                [
                  {
                    "aspectRatio": 1,
                  },
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  undefined,
                ]
              }
            >
              <View
                cacheComposition={true}
                collapsable={false}
                colorFilters={[]}
                enableMergePathsAndroidForKitKatAndAbove={false}
                forwardedRef={[Function]}
                loop={true}
                onAnimationFinish={[Function]}
                progress={0}
                resizeMode="contain"
                sourceJson="{"v":"5.7.4","fr":24,"ip":0,"op":144,"w":400,"h":400,"nm":"[Mapping] Check_lottie","ddd":0,"assets":[],"layers":[{"ddd":0,"ind":1,"ty":4,"nm":"check","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[200.221,204.483,0],"ix":2,"l":2},"a":{"a":0,"k":[0,0,0],"ix":1,"l":2},"s":{"a":1,"k":[{"i":{"x":[0.143,0.143,0.667],"y":[1,1,1]},"o":{"x":[0.333,0.333,0.333],"y":[0,0,0]},"t":6,"s":[220,220,100]},{"t":28,"s":[100,100,100]}],"ix":6,"l":2}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[0,0],[0,0],[0,0]],"o":[[0,0],[0,0],[0,0]],"v":[[36.221,-24.483],[-12.746,24.483],[-36.221,1.001]],"c":false},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"st","c":{"a":0,"k":[1,1,1,1],"ix":3},"o":{"a":0,"k":100,"ix":4},"w":{"a":0,"k":8,"ix":5},"lc":2,"lj":2,"bm":0,"nm":"Stroke 1","mn":"ADBE Vector Graphic - Stroke","hd":false},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"bm":0,"ix":1,"mn":"ADBE Vector Group","hd":false},{"ty":"tm","s":{"a":1,"k":[{"i":{"x":[0.143],"y":[1]},"o":{"x":[0.333],"y":[0]},"t":6,"s":[100]},{"t":28,"s":[0]}],"ix":1},"e":{"a":0,"k":100,"ix":2},"o":{"a":0,"k":0,"ix":3},"m":1,"ix":2,"nm":"Trim Paths 1","mn":"ADBE Vector Filter - Trim","hd":false}],"ip":6,"op":246,"st":6,"bm":0},{"ddd":0,"ind":2,"ty":4,"nm":"circulo verde","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[200,200,0],"ix":2,"l":2},"a":{"a":0,"k":[0,0,0],"ix":1,"l":2},"s":{"a":1,"k":[{"i":{"x":[0.279,0.279,0.667],"y":[1,1,1]},"o":{"x":[0.687,0.687,0.333],"y":[0,0,0]},"t":0,"s":[0,0,100]},{"i":{"x":[0.002,0.002,0.667],"y":[1,1,1]},"o":{"x":[0.221,0.221,0.333],"y":[0,0,0]},"t":14,"s":[110,110,100]},{"i":{"x":[0.002,0.002,0.667],"y":[1,1,1]},"o":{"x":[0.167,0.167,0.167],"y":[0,0,0]},"t":24,"s":[97,97,100]},{"t":33,"s":[100,100,100]}],"ix":6,"l":2}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[-66.274,0],[0,-66.274],[66.274,0],[0,66.274]],"o":[[66.274,0],[0,66.274],[-66.274,0],[0,-66.274]],"v":[[0,-120],[120,0],[0,120],[-120,0]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"st","c":{"a":0,"k":[1,1,1,1],"ix":3},"o":{"a":0,"k":100,"ix":4},"w":{"a":0,"k":7,"ix":5},"lc":1,"lj":1,"ml":4,"bm":0,"nm":"Stroke 1","mn":"ADBE Vector Graphic - Stroke","hd":false},{"ty":"fl","c":{"a":0,"k":[0.098039217293,0.858823537827,0.40000000596,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"bm":0,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":3,"cix":2,"bm":0,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":240,"st":0,"bm":0},{"ddd":0,"ind":3,"ty":4,"nm":"circulo atras 2","sr":1,"ks":{"o":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"t":18,"s":[20]},{"t":33.4501953125,"s":[0]}],"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[200,200,0],"ix":2,"l":2},"a":{"a":0,"k":[0,0,0],"ix":1,"l":2},"s":{"a":1,"k":[{"i":{"x":[0.32,0.32,0.667],"y":[1,1,1]},"o":{"x":[0.4,0.4,0.333],"y":[0,0,0]},"t":0,"s":[0,0,100]},{"t":35,"s":[120,120,100]}],"ix":6,"l":2}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[-77.32,0],[0,-77.32],[77.32,0],[0,77.32]],"o":[[77.32,0],[0,77.32],[-77.32,0],[0,-77.32]],"v":[[0,-140],[140,0],[0,140],[-140,0]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.098039217293,0.858823537827,0.40000000596,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"bm":0,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"bm":0,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":240,"st":0,"bm":0},{"ddd":0,"ind":4,"ty":4,"nm":"circulo atras","sr":1,"ks":{"o":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"t":26,"s":[20]},{"t":41.4501953125,"s":[0]}],"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[200,200,0],"ix":2,"l":2},"a":{"a":0,"k":[0,0,0],"ix":1,"l":2},"s":{"a":1,"k":[{"i":{"x":[0.32,0.32,0.667],"y":[1,1,1]},"o":{"x":[0.4,0.4,0.333],"y":[0,0,0]},"t":8,"s":[0,0,100]},{"t":43,"s":[120,120,100]}],"ix":6,"l":2}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[-77.32,0],[0,-77.32],[77.32,0],[0,77.32]],"o":[[77.32,0],[0,77.32],[-77.32,0],[0,-77.32]],"v":[[0,-140],[140,0],[0,140],[-140,0]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.098039217293,0.858823537827,0.40000000596,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"bm":0,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"bm":0,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":8,"op":248,"st":8,"bm":0},{"ddd":0,"ind":5,"ty":4,"nm":"BG","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[200,200,0],"ix":2,"l":2},"a":{"a":0,"k":[0,0,0],"ix":1,"l":2},"s":{"a":0,"k":[100,100,100],"ix":6,"l":2}},"ao":0,"shapes":[{"ty":"gr","it":[{"ty":"rc","d":1,"s":{"a":0,"k":[400,400],"ix":2},"p":{"a":0,"k":[0,0],"ix":3},"r":{"a":0,"k":0,"ix":4},"nm":"Rectangle Path 1","mn":"ADBE Vector Shape - Rect","hd":false},{"ty":"st","c":{"a":0,"k":[1,1,1,1],"ix":3},"o":{"a":0,"k":100,"ix":4},"w":{"a":0,"k":2,"ix":5},"lc":1,"lj":1,"ml":4,"bm":0,"nm":"Stroke 1","mn":"ADBE Vector Graphic - Stroke","hd":false},{"ty":"fl","c":{"a":0,"k":[1,1,1,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"bm":0,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Rectangle 1","np":3,"cix":2,"bm":0,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":240,"st":0,"bm":0}],"markers":[]}"
                speed={1}
                style={
                  {
                    "aspectRatio": 1,
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                textFiltersAndroid={[]}
                textFiltersIOS={[]}
                useNativeLooping={false}
              />
            </View>
          </View>
          <Text
            style={
              {
                "height": 12,
                "width": 12,
              }
            }
          />
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#3E4149",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 27,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 32,
                "textAlign": "center",
              }
            }
          >
            text
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
